Add Fragments and use tabs

lcc opened the pull request:
Updated the app UI and UX by adding tabs that hold the fragments that control the app workflow.

Closes #44.
Closes #45.
Closes #49.

Commit:
Update texts to new UI

Update UI texts to be more explanative and easier to understand by the
users.

Commit:
Add black line beneath the login field

Commit:
Change login to use tabs

Change login to load one tab that holds a fragment that does exactly
the same as the former login activity. This approach was taken to
prevent the use of different layouts (other than the header layout)
in order to keep the consistency of the app views.

Commit:
Change configure thing to use tabs

Change configure thing to use a single tabs in order to mantain the
header layout style.

Commit:
Change configure gateway wifi to use tabs

Change configure thing to use a single tabs in order to mantain the
header layout style.

Commit:
Change folder naming and distribution

Changed folder naming and distribution to reflect the use of mvp
architecure and therefore simplify the first contact of non-android
developers with this repository (users should be familiarized with MVP
pattern).

alla3 commented on app/src/main/java/br/org/cesar/knot_setup_app/view/ScanInterface.java:
```diff
@@ -0,0 +1,8 @@
+
+import java.util.UUID;
+
+public interface ScanInterface {
```
Is this interface needed?

lcc replied:
No, I just deleted it.

pauloserrafh replied:
@alla3 If the comment is resolved, please resolve the conversation.

alla3 commented on app/src/main/java/br/org/cesar/knot_setup_app/view/scan/ScanFragment.java:
```diff
@@ -0,0 +1,140 @@
+import br.org.cesar.knot_setup_app.R;
+import br.org.cesar.knot_setup_app.view.ConfigureDeviceActivity;
+import br.org.cesar.knot_setup_app.view.ConfigureGatewayWifiActivity;
+import br.org.cesar.knot_setup_app.view.ScanInterface;
```
You are importing this file and not using.

lcc replied:
Which file?

alla3 replied:
br.org.cesar.knot_setup_app.view.ScanInterface

lcc replied:
Removed.

alla3 commented on app/src/main/java/br/org/cesar/knot_setup_app/view/GatewayActivity.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    protected void onResume() {
```
It is necessary to override this method?

lcc replied:
This method is not necessary as a whole I will delete it now.

alla3 commented on app/src/main/java/br/org/cesar/knot_setup_app/view/LoginActivity.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @Override
+    protected void onResume() {
```
It is necessary to override this method?

lcc replied:
Same as above.

alla3 commented on app/src/main/java/br/org/cesar/knot_setup_app/view/ThingActivity.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    protected void onResume() {
```
It is necessary to override this method?

lcc replied:
Same as above.

alla3 commented on app/src/main/java/br/org/cesar/knot_setup_app/view/scan/ScanPresenter.java:
```diff
@@ -57,35 +64,40 @@ public void onScanComplete(android.bluetooth.le.ScanResult result) {
                 }
 
                 @Override
                 public void onScanFail() {
```
Is this method necessary? This method is empty and as I can see your class do not implement the interface it was declared.

lcc replied:
Not sure if I understood your statement, the class ScanFragment.java does implement the ViewModel interface that declares this method. Nevertheless, this method is used to update the list that displays the nearby Bluetooth devices (Things and Gateways).

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/configureGatewayWifi/ConfigureGatewayWifiActivity.java:
```diff
@@ -38,7 +39,7 @@ protected void onCreate(Bundle savedInstanceState) {
         sendButton = findViewById(R.id.buttonSendWifiSettings);
 
-        headerTitle.setText("Provide password");
+        headerTitle.setText(headerTitleText);
```
Define a new value on strings.xml and use `getString(R.string.<string_name>)` instead.

lcc replied:
Done.

pauloserrafh replied:
There is no change here.

lcc replied:
This was changed when changing the activity to use fragments. Would you like me to change it anyway?

pauloserrafh replied:
Yes, just as you did. The change to use the resources file should be done in this commit.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/login/LoginActivity.java:
```diff
@@ -25,7 +27,7 @@ protected void onCreate(Bundle savedInstanceState){
 
             TextView headerTitle = (TextView)findViewById(R.id.list_title);
-            headerTitle.setText("LOGIN");
+            headerTitle.setText(headerTitleText);
```
Define a new value on strings.xml and use `getString(R.string.<string_name>)` instead.

lcc replied:
Done.

pauloserrafh replied:
No change here as well

lcc replied:
Same as above.

pauloserrafh replied:
Yes, just as you did. The change to use the resources file should be done in this commit.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/GatewayActivity.java:
```diff
@@ -0,0 +1,46 @@
+
+public class GatewayActivity extends BaseActivity {
+
+    private static final String HEADER_TITLE = "Gateways";
```
If this text is being shown to the user, move it to strings.xml as it may need translation.

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/BaseActivity.java:
```diff
@@ -0,0 +1,52 @@
+package br.org.cesar.knot_setup_app.activity;
```
From what I got, this commit does more than one thing.
Please break into `creating the BaseActivity` and `adding tabs` (at least). If it's impossible to split, provide reasoning.

lcc replied:
The major problem, in my opinion, in splitting the commit is that the BaseActivity will be added but will not have any real use in the code. As far as I know, the commit should be self-contained in the sense that it should not need others commits to properly function.  This patch will change the UI of the app by adding tabs to certain activities, but in order to have tabs, I must change the activity_header to use tabs and add Tab.java and PageAdapter.java. This only adds tabs to the app but it makes no use of them and therefore does not make sense. So in the same commit, I must(?) use this tabbed view, hence changing the GatewayActivity into the fragments that are going to be used by the app tabs. Which leads to adding the ScanFragment and the ConnectedGatewayFragment and an Activity to load these fragments (which is the new GatewayActivity). 
P.S: I added the BaseActivity.java just so there is no need for every activity to implement the setupHeader method.

pauloserrafh replied:
Just as you explained, it was even more clear (IMO) that you could have a single commit creating the `BaseActivity`, as it is not related to adding the tabs at all.
It was just a wrapper to the every activity containing the `setupHeader`.

As for the commit being self-contained, it should not break the mainline for sure. I understand if the commit needs to be big as there are many changes necessary and it's OK.

lcc replied:
What do you think about using a more neutral term, such as, onInit()?

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/GatewayActivity.java:
```diff
@@ -0,0 +1,46 @@
+public class GatewayActivity extends BaseActivity {
+
+    private static final String HEADER_TITLE = "Gateways";
+    private static final String LEFT_TAB_TITLE = "Configure Wi-Fi";
```
Same as above

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/GatewayActivity.java:
```diff
@@ -0,0 +1,46 @@
+
+    private static final String HEADER_TITLE = "Gateways";
+    private static final String LEFT_TAB_TITLE = "Configure Wi-Fi";
+    private static final String RIGHT_TAB_TITLE = "Connected";
```
Same as above

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/BaseActivity.java:
```diff
@@ -0,0 +1,52 @@
+        tabLayout.setupWithViewPager(viewPager);
+    }
+
+    public void setTabLayoutVisibilityToInvisible(){
```
This method does not seem to be used anywhere. If not necessary at the moment, please remove.

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/GatewayActivity.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @Override
+    protected void onResume() {
```
If you're not overwriting onResume, it is not necessary to override it.

lcc replied:
This method is not necessary at all, it was removed.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/splash/SplashActivity.java:
```diff
@@ -36,13 +37,15 @@ protected void onCreate(Bundle savedInstanceState) {
     @Override
     public void doListGateways(){
         Intent intent = new Intent(SplashActivity.this, ThingActivity.class);
+        intent.putExtra("operation", Constants.CONFIGURE_THING_OPENTHREAD);
```
Change "operation" into a constant.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/splash/SplashActivity.java:
```diff
@@ -36,13 +37,15 @@ protected void onCreate(Bundle savedInstanceState) {
     @Override
     public void doLogin(){
         Intent intent = new Intent(SplashActivity.this, LoginActivity.class);
+        intent.putExtra(Constants.GATEWAY_ID,getIntent().getStringExtra(Constants.GATEWAY_ID));
```
IMO, if you break this would help readability. 
```
String serviceName = getIntent().getStringExtra(Constants.GATEWAY_ID);
intent.putExtra(Constants.GATEWAY_ID, serviceName);
```
This is mostly a suggestion, but it took me a second look to understand what was being done here due to all the `Constants.GATEWAY_ID` in the same line.

pauloserrafh replied:
Add space after `,`

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedFragment.java:
```diff
@@ -0,0 +1,112 @@
+
+        mView = view;
+        mPresenter = new GatewayConnectedPresenter(this,
+                (NsdManager) getActivity().getSystemService(Context.NSD_SERVICE),getContext());
```
AndroidStudio complains that `getSystemService` may produce `NullPointerException`.
Auto replace suggests: `Objects.requireNonNull(getActivity()).getSystemService(Context.NSD_SERVICE),getContext());`

Please check if this would not cause any breaking changes.

lcc replied:
No complaints on my system. It also tells me that Objects.requireNonNull() method does not exist, are you using the most recent android studio?

pauloserrafh replied:
Perhaps its something with the JDK version. I think it was introduced in JDK 7.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedFragment.java:
```diff
@@ -0,0 +1,112 @@
+        return view;
+    }
+
+    /**
```
If you're overriding a method, the javadoc should be in the initial declaration.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/ThingActivity.java:
```diff
@@ -0,0 +1,45 @@
+
+public class ThingActivity extends BaseActivity {
+
+    private static final String HEADER_TITLE = "THING's";
```
Move to strings.xml

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/ThingActivity.java:
```diff
@@ -0,0 +1,45 @@
+public class ThingActivity extends BaseActivity {
+
+    private static final String HEADER_TITLE = "THING's";
+    private static final String LEFT_TAB_TITLE = "Unregistered";
```
Move to strings.xml

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/ThingActivity.java:
```diff
@@ -0,0 +1,45 @@
+
+    private static final String HEADER_TITLE = "THING's";
+    private static final String LEFT_TAB_TITLE = "Unregistered";
+    private static final String RIGHT_TAB_TITLE = "Registered";
```
Move to strings.xml

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/thingRegistered/ThingPresenter.java:
```diff
@@ -44,8 +39,7 @@
 
-
-    public void getDeviceList(){
+    public void onResume() {
```
Why change the name here?
IMO it makes more sense to have the method called getDevicesList than onResume.

lcc replied:
It was pointed out that the view, should only tell the presenter that some event happened (e.g. a button was clicked, the view resumed, Bluetooth radio was turned off) and the view should NOT use the imperative voice to tell the presenter to do something (e.g. getSomething, setSomething, doSomething). 
As such, I opted to rename several methods to more event-related names. So for example, in this case, the activity was resuming so it seemed appropriate for the presenter method name to be warned about this particular event, hence, the name.
But given your comment about the presenter being responsible to decouple android code from the rest of the code, I think that maybe the solution would be to rename the method to something not android specific but still event-related (onViewResumed?). I would really like to hear your opinion about this.

pauloserrafh replied:
I'd go with `listDevices`, `populateView`, or something like that. I understand that the point is to tell the presenter that something happened, but the name might be confusing/too attached to android lifecycle. 
My point is along the lines of: 
Imagine if you were to change the view now, from something other than Android. The name `onResume` probably wouldn't make that much sense whereas more explicit like the ones above would.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/thingRegistered/ThingFragment.java:
```diff
@@ -0,0 +1,64 @@
+
+
+
+    /**
```
Do the javadoc in the contract. 
Doing so, if you re-define the method anywhere else, you won't need to rewrite the javadoc

lcc replied:
Sure, makes sense, but maybe(?) for a future PR.

pauloserrafh replied:
ACK.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/LoginActivity.java:
```diff
@@ -0,0 +1,36 @@
+
+public class LoginActivity extends BaseActivity {
+
+    private static final String HEADER_TITLE = "User Credentials";
```
Move to strings.xml

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/LoginActivity.java:
```diff
@@ -0,0 +1,36 @@
+            setupHeader(HEADER_TITLE,tabs);
+    }
+
+    @Override
```
If only calling the super method, no need to have this overriden

lcc replied:
The method was not necessary, removed.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginContract.java:
```diff
@@ -0,0 +1,15 @@
+
+    interface ViewModel {
+        void fillEmail(String email);
+        void callbakcOninvalidCredentials();
```
The `on` prefix already denotes a callback. No need to call it `callbackOnSomething`.

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginContract.java:
```diff
@@ -0,0 +1,15 @@
+    }
+
+    interface Presenter {
+        void onResume();
```
IMO the presenter should not have a method name onResume. onResume is an Android Lifecycle specific method and the idea of the presenter is to decouple Android specific code from the rest of the code.
Just name the method after what it does (or have multiple methods if necessary).

lcc replied:
This discussion is happening in another comment.

pauloserrafh reviewed (changes requested):
Some points of the review up to now:
- Break the commits into smaller and feature specific commits (the fact that you are resolving 3+ issues in only one PR already points to something weird). There is a lot of refactoring being done in this patchset, break it accordingly.
- Kudos on having javadocs in (some) methods. I'd suggest having in all of them.
- I got the idea of reusing layout files, but sometimes it might be weird. If I understood correctly the idea of 04331d178eee6a328cee7558054b21fdab7ed817 , you have one activity with one fragment with the sole purpose of reusing the layout. If it is not possible to re-use the same layout without having to create a fragment, I'd go with creating a new one. AFAIK the idea of fragments is to have only a small portion of the screen changing, but this seems like overstretching to use it as you're doing (or from what I understood you're doing).

lcc committed:
Add gateway activity that holds two tabs

Add gateway activty that holds two fragments. Also added a base activity
since other activities (ThingActivity and so on) will need to use tabs.
BaseActivity is used to  abstract the use of tabs.

lcc committed:
Add gatewayConnected fragment

Add gatewayConnected fragment that finds, through mDNS, all the gateways
that are connected to the same Wi-Fi as the app.

lcc committed:
Add scanFragment

Add scanFragment that searches for gateways that are advertising
somekind of service (the service is specified by the calling activity).

pauloserrafh commented on app/src/main/res/layout/activity_configure_gateway_wifi.xml:
```diff
@@ -43,7 +43,7 @@
         android:layout_marginStart="8dp"
         android:layout_marginTop="8dp"
-        android:text="Password"
+        android:text="@string/password"
```
Notice that `@string/password` contains `:`, whereas the hard-coded one did not. Check if it's OK to have the colon or if the string is used somewhere else and it is fine to remove it.

lcc replied:
It's alright, the colon is correct. This string is only used by fragment_login.xml.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedPresenter.java:
```diff
@@ -0,0 +1,165 @@
+        //stores the ip in the user preferences
+        dataManager.getInstance().getPreference().
+                setSharedPreferenceString(context,
+                        "ip",serviceInfo.getHost().getHostAddress());
```
Move `"ip"` this to a constant.

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedPresenter.java:
```diff
@@ -0,0 +1,165 @@
+        //stores the port in the user preferences
+        dataManager.getInstance().getPreference().
+                setSharedPreferenceString(context,
+                        "port", String.valueOf(serviceInfo.getPort()));
```
Move `"port"` this to a constant.

lcc replied:
Done

pauloserrafh replied:
Done in the previous commit and re-introduced here.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/scan/ScanFragment.java:
```diff
@@ -0,0 +1,139 @@
+        mHandler.post(new Runnable() {
+            @Override
+            public void run() {
+                Toast.makeText(getActivity(), "Scan for device" +
```
Move this to the string resources file.

lcc replied:
Done

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -7,6 +7,11 @@
 
+    <!-- Strings used by ThingActivity -->
+    <string name="thing_header_title">THING\'s</string>
+    <string name="thing_left_tab_name">Unregistered</string>
```
If you name based on 'left_tab' or 'right_tab' you're kind of stuck with this positioning.
Why not `thing_unregistered_tab` and `thing_registered_tab` or something like that?
Please note that the same applies to the gateway patch.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -1,28 +1,28 @@
     private String request,port,ip;
     private Context context;
 
+
```
Remove this line

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -33,32 +33,40 @@
                 .getPreference().getSharedPreferenceString(context,"ip");
 
         this.port = dataManager.getInstance()
-                    .getPreference().getSharedPreferenceString(context,"port");
```
This tabulation should be manteined. Perhaps better, should be increased to 8 spaces.
If the entire line after `=`fits in 100 characters, move the entire line down.
https://source.android.com/setup/contribute/code-style#use-spaces-for-indentation

pauloserrafh replied:
Still an issue.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginFragment.java:
```diff
@@ -0,0 +1,77 @@
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.fragment_login, container, false);
+
+        emailFld = view.findViewById(R.id.email);
```
There is no point in shortening `Field` to `Fld`. Only decreases readability.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginFragment.java:
```diff
@@ -0,0 +1,77 @@
+        View view = inflater.inflate(R.layout.fragment_login, container, false);
+
+        emailFld = view.findViewById(R.id.email);
+        pwdFld =  view.findViewById(R.id.password_input);
```
Same as above.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginFragment.java:
```diff
@@ -0,0 +1,77 @@
+        mHandler.post(new Runnable() {
+            @Override
+            public void run() {
+                Toast.makeText(getContext(),"Incorrect username or password."
```
Move this to strings resource file

pauloserrafh replied:
Add space after `,`

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginFragment.java:
```diff
@@ -0,0 +1,77 @@
+        button.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                presenter.onSendClicked(emailFld.getText().toString(),pwdFld.getText().toString());
```
add space after `,`

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginFragment.java:
```diff
@@ -0,0 +1,77 @@
+    private EditText emailFld;
+    private EditText pwdFld;
+    private Presenter presenter;
+    private Handler mHandler;
```
Why use hungarian notation only for one variable? Choose one or another.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureGatewayWifi/ConfigureGatewayWifiContract.java:
```diff
@@ -0,0 +1,15 @@
+
+    interface Presenter {
+        void onFocus();
+        void onSetWifiClicked(String ssid, String pssd);
```
if `pssd` means password, either use the entire word or keep the same pattern as the previous commit (I think it's pwd).

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureGatewayWifi/ConfigureGatewayWifiFragment.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public void onWriteFailed() {
+        sendButton.setEnabled(true);
+        Toast.makeText(getActivity(),"Failure when writing characteristic",Toast.LENGTH_SHORT).show();
```
Move the code to the strings resource file.
Add space after `,`

lcc replied:
Wouldn't it make more sense to change this in another PR focused on coding style/readability? (change all toasts for example)

pauloserrafh replied:
No, you are adding the toast in this commit. Also, the text should be in the resources file.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureGatewayWifi/ConfigureGatewayWifiPresenter.java:
```diff
@@ -1,16 +1,16 @@
 
-public class ConfigureGatewayWifiPresenter  implements Presenter {
-
+public class ConfigureGatewayWifiPresenter implements Presenter{
```
Why change this line?
You are fixing one coding style issue (double spacing) and adding another (no space before `{`).
As discussed, I'd suggest fix all this coding style issues in a final commit in this patchset, since they are already in the mainline.

lcc replied:
Done.

pauloserrafh replied:
Not done. You are still changing the code style in an incorrect patch.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureGatewayWifi/ConfigureGatewayWifiPresenter.java:
```diff
@@ -97,7 +101,7 @@ public void onCharacteristicChanged() {
     }
 
-    public void gatewayWifiConfigWrite(String value){
+    public void gatewayWifiConfigWrite(String value) {
```
Same as in the class declaration.

lcc replied:
Done.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureGatewayWifi/ConfigureGatewayWifiFragment.java:
```diff
@@ -0,0 +1,79 @@
+
+public class ConfigureGatewayWifiFragment extends Fragment implements ViewModel {
+
+    private EditText wifiEditText;
```
In the previous commit I think you used `somethingFld` (which should be replaced by Field). But try to keep a standard.

lcc replied:
I decided to adopt this pattern <variable name><XmlField> in order to keep a clear distinction from text fields to edit texts, I would like to hear your thoughts about this. Also, are you requesting any changes? I couldn't tell from the comment.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureGatewayWifi/ConfigureGatewayWifiFragment.java:
```diff
@@ -0,0 +1,79 @@
+public class ConfigureGatewayWifiFragment extends Fragment implements ViewModel {
+
+    private EditText wifiEditText;
+    private EditText passwEditText;
```
passw, pwd, pssd.

lcc replied:
Are you requesting any changes?

pauloserrafh reviewed (changes requested):
There are automatically code style issues raised by github.

As discussed, keep the coding style changes in a separate commit.
Suggest using a lint to verify coding style issues:
https://developer.android.com/studio/write/lint

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -29,6 +29,13 @@
+    <string name="loginSubtitle">User Credentials</string>
 
+    <!-- Strings used in activity_configure_gateway_wifi -->
+    <string name="networkName">Network name:</string>
```
For `.xml`files, I'd suggest we go with snake case for variable naming.
See this post from a GDE for a suggestion on how to name variables: https://jeroenmols.com/blog/2016/03/07/resourcenaming/

Also, looking at android documentation, I think it's safe to assume that for xml files its more common to go with snake_case than with camelCase.
https://developer.android.com/guide/topics/resources/string-resource

pauloserrafh commented on app/src/main/res/layout/fragment_list.xml:
```diff
@@ -0,0 +1,47 @@
+        />
+
+    <TextView
+        android:id="@+id/searchingForGatewayText"
```
see comment on first commit

pauloserrafh commented on app/src/main/res/layout/fragment_list.xml:
```diff
@@ -0,0 +1,47 @@
+        />
+
+    <ProgressBar
+        android:id="@+id/progressBar3"
```
see comment on first commit.
Also, why is it `progressBar3`? Are there other progress bars in this layout? Rename to something more descriptive.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/GatewayActivity.java:
```diff
@@ -0,0 +1,30 @@
+        tabs.add(leftTab);
+        tabs.add(rightTab);
+
+        setupHeader("Gateways", tabs);
```
Move this to the resources file.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/PageAdapter.java:
```diff
@@ -0,0 +1,37 @@
+
+public class PageAdapter extends FragmentPagerAdapter {
+
+    private ArrayList<Fragment> mFragmentList = new ArrayList<>();
```
Lets avoid using hungarian notation and try to stick to google style. Remove the `m` prefix.

https://google.github.io/styleguide/javaguide.html#s5-naming
https://blog.mindorks.com/android-code-style-and-guidelines-d5f80453d5c7

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/GatewayActivity.java:
```diff
@@ -17,14 +20,14 @@ protected void onCreate(Bundle savedInstanceState) {
         tabs.add(rightTab);
 
-        setupHeader("Gateways", tabs);
+        setupHeader(getString(R.string.gateway_header_title), tabs);
```
Do this change in the previous commit where you create the activity.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/GatewayActivity.java:
```diff
@@ -17,14 +20,14 @@ protected void onCreate(Bundle savedInstanceState) {
-        leftTab = new Tab("left-tab", new Fragment());
-        rightTab = new Tab("right-tab",
-                new Fragment());
+        leftTab = new Tab(getString(R.string.gatewayConfigureWiFi), new Fragment());
```
Now that the tabs are not just dummy values, rename the variables to something more descriptive and not attached to their positioning as we talked before.
IMO, it's OK to have that name in the previous commit as there were not a better naming at the moment.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedFragment.java:
```diff
@@ -0,0 +1,112 @@
+    private Presenter mPresenter;
+    private ArrayList<NsdServiceInfo> deviceList;
+    private GatewayAdapter adapter;
+    private Handler mHandler;
```
See previous comment about variable naming.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedFragment.java:
```diff
@@ -0,0 +1,112 @@
+
+        mView = view;
+        mPresenter = new GatewayConnectedPresenter(this,
+                (NsdManager) getActivity().getSystemService(Context.NSD_SERVICE),getContext());
```
Add space after `,`

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedFragment.java:
```diff
@@ -0,0 +1,112 @@
+                mPresenter.onGatewayClicked(deviceList.get(position));
+                LogWrapper.Log("Service name: " + deviceList.get(position).getServiceName(), Log.DEBUG);
+                Intent i = new Intent(getActivity(), SplashActivity.class);
+                i.putExtra(Constants.GATEWAY_ID,deviceList.get(position).getServiceName());
```
Add space after `,`

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedPresenter.java:
```diff
@@ -0,0 +1,165 @@
+    private ArrayList<NsdServiceInfo> mService;
+
+    GatewayConnectedPresenter(ViewModel viewModel, NsdManager nsdManager, Context context) {
+        mViewModel = viewModel;
```
See previous comments about variable naming

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedPresenter.java:
```diff
@@ -0,0 +1,165 @@
+    private NsdManager.DiscoveryListener discoveryListener;
+    private static DataManager dataManager;
+
+    private ArrayList<NsdServiceInfo> mService;
```
See previous comments about variable naming

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedPresenter.java:
```diff
@@ -0,0 +1,165 @@
+
+
+    public void stopScanning() {
+        LogWrapper.Log("Service discovery stopped.",Log.DEBUG);
```
Add space after `,`

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedPresenter.java:
```diff
@@ -0,0 +1,165 @@
+        if(mService.isEmpty()) {
+            mViewModel.setSearchingFeedback(View.INVISIBLE);
+        }
+
```
Remove this empty line.
`else` clause should start in the same line as closing bracket.

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,17 +1,11 @@
-    <string name="list_gateway_title">KNoT</string>
+    <!-- Strings used by ConnectedGatewayActivity -->
+    <string name="gateway_header_title">Gateways</string>
+    <string name="gatewayConfigureWiFi">Configure Wi-Fi</string>
```
See previous comments about variable naming

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,17 +1,11 @@
+    <!-- Strings used by ConnectedGatewayActivity -->
+    <string name="gateway_header_title">Gateways</string>
+    <string name="gatewayConfigureWiFi">Configure Wi-Fi</string>
+    <string name="gatewayConnected">Connected</string>
```
See previous comments about variable naming

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/scan/ScanFragment.java:
```diff
@@ -0,0 +1,138 @@
+        adapter = new ScanAdapter(getActivity(), R.layout.item_device, deviceList);
+        deviceListView = view.findViewById(R.id.list);
+
+        presenter = new ScanPresenter(this,service);
```
Add space after `,`

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/scan/ScanFragment.java:
```diff
@@ -0,0 +1,138 @@
+    private Presenter presenter;
+    private List<BluetoothDevice> deviceList;
+    private UUID service;
+    private Handler mHandler;
```
See previous comments about variable naming.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/scan/ScanPresenter.java:
```diff
@@ -0,0 +1,103 @@
+
+        if(service == Constants.WIFI_CONFIGURATION_SERVICE_GATEWAY) {
+            viewModel.onGatewayWifiConfiguration();
+        }else if (service == Constants.OT_SETTINGS_SERVICE) {
```
add space after `}`

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -31,5 +31,5 @@
     <string name="buttonEnter">enter</string>
-
+    <string name="scanFailed">Scan for device has failed. Check if your bluetooth is available.</string>
 </resources>
```
See previous commits about variable naming

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/ThingActivity.java:
```diff
@@ -0,0 +1,38 @@
+        scanFragment.setService(Constants.OT_SETTINGS_SERVICE);
+
+        Tab leftTab, rightTab;
+        leftTab = new Tab(getString(R.string.thingUnregistered), scanFragment);
```
As discussed, rename this to avoid references to placement of the tab.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/activity/ThingActivity.java:
```diff
@@ -0,0 +1,38 @@
+        tabs.add(leftTab);
+        tabs.add(rightTab);
+
+        setupHeader(getString(R.string.thing_header_title),tabs);
```
Add space after `,`.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/thingRegistered/ThingPresenter.java:
```diff
@@ -1,51 +1,46 @@
 import io.reactivex.schedulers.Schedulers;
 
-public class  ThingPresenter implements Presenter{
+public class ThingPresenter implements Presenter{
```
do not make coding style changes in this patchset. As discussed, this will be fixed in a future PR.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/thingRegistered/ThingPresenter.java:
```diff
@@ -1,51 +1,46 @@
         this.token = dataManager.getInstance()
                 .getPersistentPreference()
-                .getSharedPreferenceString(context,"token");
+                .getSharedPreferenceString(context, Constants.GATEWAY_TOKEN);
```
do not make coding style changes in this patchset. As discussed, this will be fixed in a future PR.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/thingRegistered/ThingPresenter.java:
```diff
@@ -1,51 +1,46 @@
         this.ip = dataManager.getInstance()
                 .getPreference()
-                .getSharedPreferenceString(context,"ip");
+                .getSharedPreferenceString(context, Constants.GATEWAY_IP);
```
do not make coding style changes in this patchset. As discussed, this will be fixed in a future PR.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/thingRegistered/ThingPresenter.java:
```diff
@@ -1,51 +1,46 @@
 
         this.port = dataManager.getInstance()
-                .getPreference().getSharedPreferenceString(context,"port");
+                .getPreference().getSharedPreferenceString(context, Constants.GATEWAY_PORT);
```
do not make coding style changes in this patchset. As discussed, this will be fixed in a future PR.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/utils/Constants.java:
```diff
@@ -52,6 +52,7 @@
     //USER PREFERENCES KEYS
     public static final String GATEWAY_IP = "ip";
     public static final String GATEWAY_PORT = "port";
+    public static final String GATEWAY_TOKEN = "token";
```
do not make coding style changes in this patchset. As discussed, this will be fixed in a future PR.

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -7,6 +7,11 @@
 
+    <!-- Strings used by ThingActivity -->
+    <string name="thing_header_title">THING\'s</string>
+    <string name="thingUnregistered">Unregistered</string>
```
See previous comments about variable naming

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -7,6 +7,11 @@
+    <!-- Strings used by ThingActivity -->
+    <string name="thing_header_title">THING\'s</string>
+    <string name="thingUnregistered">Unregistered</string>
+    <string name="thingRegistered">Registered</string>
```
See previous comments about variable naming

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
     private void loginSucceeded(User user) {
         dataManager.getInstance().getPersistentPreference()
-                .setSharedPreferenceString(context,"email",email);
+                .setSharedPreferenceString(context, Constants.GATEWAY_USER_EMAIL, email);
```
do not make coding style changes in this patchset. As discussed, this will be fixed in a future PR.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
 
         dataManager.getInstance().getPersistentPreference()
-                .setSharedPreferenceString(context,"token","Bearer " + user.getToken());
+                .setSharedPreferenceString(context, Constants.GATEWAY_TOKEN, "Bearer " + user.getToken());
```
do not make coding style changes in this patchset. As discussed, this will be fixed in a future PR.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
-        .observeOn(AndroidSchedulers.mainThread())
-        .subscribe(this::loginSucceeded,
-        this::onErrorHandler);
+                .timeout(30, TimeUnit.SECONDS)
```
do not make coding style changes in this patchset. As discussed, this will be fixed in a future PR.

pauloserrafh commented on app/src/main/res/values/dimens.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <dimen name="textSizeSubTittle">16sp</dimen>
```
See previous comments about variable naming.

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -21,10 +21,12 @@
     <string name="login">Login:</string>
     <string name="password">Password:</string>
     <string name="loginSubtitle">User Credentials</string>
+    <string name="invalidCredentials">Incorrect username or password.</string>
```
See previous comments about variable naming.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDeviceFragment.java:
```diff
@@ -0,0 +1,172 @@
+        View view = inflater.inflate(R.layout.fragment_configure_thing, container, false);
+        mView = view;
+
+        int gatewayID = getActivity().getIntent().getIntExtra("gatewayID",0);
```
Move both values to constants

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDeviceFragment.java:
```diff
@@ -0,0 +1,172 @@
+        mView = view;
+
+        int gatewayID = getActivity().getIntent().getIntExtra("gatewayID",0);
+        int operation = getActivity().getIntent().getIntExtra("operation",
```
Move operation to constants

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDeviceFragment.java:
```diff
@@ -0,0 +1,172 @@
+        int operation = getActivity().getIntent().getIntExtra("operation",
+                Constants.CONFIGURE_THING_OPENTHREAD);
+
+        this.presenter = new ConfigureDevicePresenter(this,gatewayID, operation
```
Add space after `,`

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDeviceFragment.java:
```diff
@@ -0,0 +1,172 @@
+        mHandler.post(new Runnable() {
+            @Override
+            public void run() {
+                Toast.makeText(getActivity().getApplicationContext(), "Successfully paired" +
```
Move text to resources file

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDeviceFragment.java:
```diff
@@ -0,0 +1,172 @@
+            @Override
+            public void run() {Toast.makeText(getActivity().getApplicationContext(),
+                    "Something is wrong with the gateway, please try again later.",
+                    Toast.LENGTH_LONG).show();
```
Move text to resources file

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDeviceFragment.java:
```diff
@@ -0,0 +1,172 @@
+            public void run() {
+
+                switch (val){
+                    case 0:
```
Use constants here

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDeviceFragment.java:
```diff
@@ -0,0 +1,172 @@
+            @Override
+            public void run() {
+
+                switch (val){
```
Isn't it possible to create an enumerator for this values?

pauloserrafh replied:
What about this?

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -39,14 +40,18 @@
     private static DataManager dataManager;
 
-    ConfigureDevicePresenter(ViewModel viewModel,int gatewayID, int operation,Context context){
+    ConfigureDevicePresenter(ViewModel viewModel, int gatewayID, int operation, Context context){
```
Do not make style changes in this PR

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -141,17 +147,17 @@ public void onCharacteristicChanged(){
             public void onServiceDiscoveryComplete(){
                 LogWrapper.Log("Services discovered", Log.DEBUG);
-                if(operation == Constants.CONFIGURE_THING_OPENTHREAD){
+                if(operation == Constants.CONFIGURE_THING_OPENTHREAD) {
```
Do not make style changes in this PR

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -141,17 +147,17 @@ public void onCharacteristicChanged(){
                 }
                 //Todo: check if the logic of this else makes sense
-                else{
+                else {
```
Do not make style changes in this PR

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -163,11 +169,11 @@ public void onServiceDiscoveryFail(){
+                mViewModel.onWriteSucceeded(writeCount);
 
-                if(writeDone){
+                if(writeDone) {
```
Do not make style changes in this PR

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -177,9 +183,9 @@ public void onCharacteristicWriteComplete(){
 
             @Override
-            public void onCharacteristicWriteFail(){
+            public void onCharacteristicWriteFail() {
```
Do not make style changes in this PR

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -201,9 +207,9 @@ public void onCharacteristicReadComplete(byte[] value){
+                mViewModel.onWriteSucceeded(readCount);
 
-                if(readDone){
+                if(readDone) {
```
Do not make style changes in this PR

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -12,7 +12,9 @@
-    <!--Strings used in configure_gateway_activity -->
+    <!--Strings used by ConfigureDeviceActivity -->
+    <string name="configure_device_header_title">Network Settings</string>
+    <string name="configure_device_tab_name">KNoT gateway id?</string>
```
Is the question mark correct here?

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/view/BaseActivity.java:
```diff
@@ -11,7 +11,6 @@
 import java.util.ArrayList;
 
 import br.org.cesar.knot_setup_app.R;
-import br.org.cesar.knot_setup_app.fragment.PageAdapter;
```
If this import is not necessary, remove in the original commit.

pauloserrafh replied:
From what I understood, this is still being used, why remove?

pauloserrafh reviewed (changes requested):
- Please rewrite commit message for bc3b9abd3f00721bacd54d29bf307d965de0cc20.
Do not mention activities that were not created yet, also, you’re saying `Add gateway activity that holds two tabs` and `Add gateway activty that holds two fragments.`  I understand what you meant, but it’s not a clear message (you’d have to look into the code, the message is not descriptive enough).

- commit message c643fe1232bae90aee9e302ae38ee2be3f21f0ec `trough` is a typo? Also, the header and first line are the same. I suggest something like:
```
Add gatewayConnected fragment

Use mDNS to find all KNoT gateways in the same network as the KNoT app.  
```

- commit message afb99b7c2e12aaeb652bb5b4f511178d5034a7d0 `somekind`is a typo? Also, it could be rewritten.

lcc committed:
Add tabs to thing list

Add tabs to thing list by creating a fragment to list the things and
reutilizing the scan fragment.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/scan/ScanPresenter.java:
```diff
@@ -0,0 +1,103 @@
+                            break;
+                        }
+                    }
+                    deviceList.add(new BluetoothDevice(result.getDevice(),result.getRssi()));
```
Add space after `,`

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/gatewayConnected/GatewayConnectedPresenter.java:
```diff
@@ -61,11 +61,11 @@ public void onGatewayClicked(NsdServiceInfo serviceInfo) {
         dataManager.getInstance().getPreference().
                 setSharedPreferenceString(context,
-                        Constants.GATEWAY_IP, serviceInfo.getHost().getHostAddress());
+                        "ip", serviceInfo.getHost().getHostAddress());
```
Undo this.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/thingRegistered/ThingPresenter.java:
```diff
@@ -1,51 +1,46 @@
         this.token = dataManager.getInstance()
                 .getPersistentPreference()
-                .getSharedPreferenceString(context,"token");
+                .getSharedPreferenceString(context, "token");
```
This space insertion should not be done in this commit.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/thingRegistered/ThingPresenter.java:
```diff
@@ -1,51 +1,46 @@
         this.ip = dataManager.getInstance()
                 .getPreference()
-                .getSharedPreferenceString(context,"ip");
+                .getSharedPreferenceString(context, "ip");
```
This space insertion should not be done in this commit.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/thingRegistered/ThingPresenter.java:
```diff
@@ -1,51 +1,46 @@
 
         this.port = dataManager.getInstance()
-                .getPreference().getSharedPreferenceString(context,"port");
+                .getPreference().getSharedPreferenceString(context, "port");
```
This space insertion should not be done in this commit.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/utils/Constants.java:
```diff
@@ -49,10 +49,6 @@
     public static final String GATEWAY_ID = "GATEWAY_ID";
 
-    //USER PREFERENCES KEYS
-    public static final String GATEWAY_IP = "ip";
```
Why is this being removed?

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/utils/Constants.java:
```diff
@@ -49,10 +49,6 @@
 
-    //USER PREFERENCES KEYS
-    public static final String GATEWAY_IP = "ip";
-    public static final String GATEWAY_PORT = "port";
```
Why is this being removed?

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
 
         this.ip = dataManager.getInstance()
-                .getPreference().getSharedPreferenceString(context,"ip");
+                .getPreference().getSharedPreferenceString(context, "ip");
```
This space insertion should not be done in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
-        .observeOn(AndroidSchedulers.mainThread())
-        .subscribe(this::loginSucceeded,
-        this::onErrorHandler);
+                .timeout(30, TimeUnit.SECONDS)
```
This space insertion should not be done in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
-        .subscribe(this::loginSucceeded,
-        this::onErrorHandler);
+                .timeout(30, TimeUnit.SECONDS)
+                .subscribeOn(Schedulers.newThread())
```
This space insertion should not be done in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
-        this::onErrorHandler);
+                .timeout(30, TimeUnit.SECONDS)
+                .subscribeOn(Schedulers.newThread())
+                .observeOn(AndroidSchedulers.mainThread())
```
This space insertion should not be done in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
+                .timeout(30, TimeUnit.SECONDS)
+                .subscribeOn(Schedulers.newThread())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(this::loginSucceeded,
```
This space insertion should not be done in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
+                .subscribeOn(Schedulers.newThread())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(this::loginSucceeded,
+                        this::onErrorHandler);
```
This space insertion should not be done in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
     private void loginSucceeded(User user) {
         dataManager.getInstance().getPersistentPreference()
-                .setSharedPreferenceString(context,"email",email);
+                .setSharedPreferenceString(context, "email", email);
```
This space insertion should not be done in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/login/LoginPresenter.java:
```diff
@@ -30,51 +30,59 @@
 
         dataManager.getInstance().getPersistentPreference()
-                .setSharedPreferenceString(context,"token","Bearer " + user.getToken());
+                .setSharedPreferenceString(context, "token", "Bearer " + user.getToken());
```
This space insertion should not be done in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDeviceFragment.java:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    @Override
+    public void onWriteFailed(int val) {
```
Same comments as onWriteSucceeded

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDeviceFragment.java:
```diff
@@ -0,0 +1,176 @@
+                        addCheck("channel");
+                        break;
+                    case 1:
+                        removeProgressBar("net_name");
```
Create constants for the string values

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -1,27 +1,28 @@
-import br.org.cesar.knot_setup_app.wrapper.BluetoothWrapper;
 import br.org.cesar.knot_setup_app.utils.Constants;
+import br.org.cesar.knot_setup_app.wrapper.BluetoothWrapper;
 import br.org.cesar.knot_setup_app.wrapper.LogWrapper;
```
Why are you moving the placement of the import?

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -1,27 +1,28 @@
 
-
 public class ConfigureDevicePresenter implements Presenter{
+
```
Remove this empty line.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -232,15 +238,15 @@ private void writeWrapper(UUID service, UUID characteristic, byte[] valtoWrite){
     }
 
-    private void readWrapper(UUID service, UUID characteristic){
+    private void readWrapper(UUID service, UUID characteristic) {
```
Do not add this space in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -232,15 +238,15 @@ private void writeWrapper(UUID service, UUID characteristic, byte[] valtoWrite){
     }
 
-    private void thingConfigWrite(){
+    private void thingConfigWrite() {
```
Do not add this space in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureDevice/ConfigureDevicePresenter.java:
```diff
@@ -232,15 +238,15 @@ private void writeWrapper(UUID service, UUID characteristic, byte[] valtoWrite){
         byte[] valueBytes;
 
-        switch (writeCount){
+        switch (writeCount) {
```
Do not add this space in this commit

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureGatewayWifi/ConfigureGatewayWifiFragment.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public void onWriteFailed() {
+        sendButton.setEnabled(true);
+        Toast.makeText(getActivity(),"Failure when writing characteristic", Toast.LENGTH_SHORT).show();
```
Move this to a constant.

pauloserrafh commented on app/src/main/java/br/org/cesar/knot_setup_app/fragment/configureGatewayWifi/ConfigureGatewayWifiPresenter.java:
```diff
@@ -1,16 +1,16 @@
 
-public class ConfigureGatewayWifiPresenter  implements Presenter {
-
+public class ConfigureGatewayWifiPresenter  implements Presenter{
```
Do not make style changes in this commit.

pauloserrafh commented on app/src/main/res/values/strings.xml:
```diff
@@ -41,10 +41,13 @@
     <string name="network_name">Network name:</string>
     <string name="configure_gateway_subtitle">Wi-Fi Configuration</string>
 
+
```
Remove this empty line
